File: analysis/flow_state_step1.R
#!/usr/bin/env Rscript
# flow_state_step1.R
# Provenance: generating script for analysis/flow_state_step1_R.csv
# Step-1 Label-Only Diagnostic (blind R implementation).
# Spec: analysis/flow-state-step1-diagnostic-spec.md
#
# Implemented independently from spec + raw data only (blind to Python).
# Deterministic; all timestamps parsed as UTC.

suppressWarnings(suppressMessages({
  options(stringsAsFactors = FALSE)
}))

ANALYSIS_DIR <- "/Users/sebjilke/Desktop/PotomacPulse/analysis"
SERIES_PATH  <- file.path(ANALYSIS_DIR, "hourly_backtest_data_v361.csv")
RESID_PATH   <- file.path(ANALYSIS_DIR, "ci_residuals_v361_multi.csv")
OUT_PATH     <- file.path(ANALYSIS_DIR, "flow_state_step1_R.csv")

# ---------------------------------------------------------------------------
# 1. Load series; parse timestamps UTC; drop blank por_now; sort ascending.
# ---------------------------------------------------------------------------
series <- read.csv(SERIES_PATH, colClasses = "character", check.names = FALSE)

# timestamp format "YYYY-MM-DD HH:MM" (UTC)
series_ts <- as.POSIXct(series$timestamp, format = "%Y-%m-%d %H:%M", tz = "UTC")
por_now   <- suppressWarnings(as.numeric(series$por_now))

# Drop rows with blank/NA por_now or unparseable timestamp
keep <- !is.na(por_now) & !is.na(series_ts)
series_ts <- series_ts[keep]
por_now   <- por_now[keep]

# Sort ascending by timestamp (stable)
ord       <- order(series_ts)
series_ts <- series_ts[ord]
por_now   <- por_now[ord]

# Numeric timestamp vector (seconds since epoch) for fast findInterval lookups.
series_tnum <- as.numeric(series_ts)
n_series    <- length(series_tnum)

cat(sprintf("Series rows after dropping blank por_now: %d\n", n_series))

# ---------------------------------------------------------------------------
# 2. flowStateAsOf classifier (vectorized over many refTimes).
#    For each refTime t:
#      hist = rows with ts <= t. If count(hist) < 8 -> 'steady'
#      current = por_now at largest ts <= t  (nearest on-or-before)
#      past    = por_now at largest ts <= t - 6h (by TIME)
#      if past is None -> 'steady'
#      change = current - past
#      thr = max(100, 0.02*current)
#      if abs(change) >= thr: 'rising' if change>0 else 'falling'
#      else 'steady'
#
#  findInterval(x, vec): largest index i with vec[i] <= x (vec sorted ascending).
#  count(ts <= t) == that index (since vec sorted ascending, index = #(<= t)).
# ---------------------------------------------------------------------------
SIX_H_SECS <- 6 * 3600

flowStateAsOf_vec <- function(ref_tnum) {
  # ref_tnum: numeric vector of reference times (seconds since epoch, UTC).
  out <- rep("steady", length(ref_tnum))

  # idx_cur: largest series index with series_tnum <= ref. Also == # history rows (hist count).
  idx_cur <- findInterval(ref_tnum, series_tnum)
  # idx_past: largest series index with series_tnum <= ref - 6h
  idx_past <- findInterval(ref_tnum - SIX_H_SECS, series_tnum)

  # History gate: hist count (= idx_cur) must be >= 8. If < 8 -> steady (already default).
  # Also need idx_cur >= 1 (a current row exists on-or-before). If idx_cur==0, no current
  # -> hist count 0 < 8 -> steady. Covered by the >=8 gate.
  ok_hist <- idx_cur >= 8

  # past is None when idx_past == 0 (no row on-or-before ref-6h). -> steady.
  ok_past <- idx_past >= 1

  active <- ok_hist & ok_past
  if (any(active)) {
    ci <- idx_cur[active]
    pi <- idx_past[active]
    cur  <- por_now[ci]
    past <- por_now[pi]
    change <- cur - past
    thr <- pmax(100, 0.02 * cur)
    st <- rep("steady", length(change))
    rising  <- (abs(change) >= thr) & (change > 0)
    falling <- (abs(change) >= thr) & (change <= 0)  # change>0 -> rising else falling; ties (==0) cannot reach thr>=100
    st[rising]  <- "rising"
    st[falling] <- "falling"
    out[active] <- st
  }
  out
}

# ---------------------------------------------------------------------------
# 3. Load predictions; parse predTs UTC; build arms.
# ---------------------------------------------------------------------------
pred <- read.csv(RESID_PATH, colClasses = "character", check.names = FALSE)

# predTs ISO UTC "2011-12-01T05:00:00.000Z"
predTs <- as.POSIXct(pred$predTs, format = "%Y-%m-%dT%H:%M:%OS", tz = "UTC")
stopifnot(!any(is.na(predTs)))

rawResidual   <- as.numeric(pred$rawResidual)
flowBin       <- pred$flowBin
flowState_log <- pred$flowState
predTs_num    <- as.numeric(predTs)

N_pred <- length(predTs_num)
cat(sprintf("Prediction rows: %d\n", N_pred))

# travel_time_h for Parcel_model: per the join, look up travel_time_h on the series at
# predTs (floor to hour join). The series travel_time_h column is per-hour PoR->GF lag.
# We join predTs (floored to UTC hour) to series timestamp. But the SERIES we kept dropped
# blank por_now rows; travel_time lookup should use the original full series travel_time.
# Spec: Parcel_model = flowStateAsOf(series, predTs - travel_time_h(predTs) h).
# travel_time_h(predTs): the travel_time at the prediction's hour.
series_full <- read.csv(SERIES_PATH, colClasses = "character", check.names = FALSE)
sf_ts  <- as.POSIXct(series_full$timestamp, format = "%Y-%m-%d %H:%M", tz = "UTC")
sf_tt  <- suppressWarnings(as.numeric(series_full$travel_time_h))
sf_tnum <- as.numeric(sf_ts)

# Floor predTs to hour for the join (predTs already on the hour, but be safe).
pred_hour_num <- floor(predTs_num / 3600) * 3600
sf_hour_num   <- floor(sf_tnum / 3600) * 3600
# Map: hour -> travel_time. Use match on floored hour.
tt_map_idx <- match(pred_hour_num, sf_hour_num)
travel_tt  <- sf_tt[tt_map_idx]
cat(sprintf("Parcel_model: predTs hours matched to series travel_time: %d / %d (NA tt: %d)\n",
            sum(!is.na(tt_map_idx)), N_pred, sum(is.na(travel_tt))))

# Build arm reference times.
arm_specs <- list(
  list(arm = "A_current",   lag_h = 0,  ref = predTs_num),
  list(arm = "Parcel_L6",   lag_h = 6,  ref = predTs_num - 6 * 3600),
  list(arm = "Parcel_L12",  lag_h = 12, ref = predTs_num - 12 * 3600),
  list(arm = "Parcel_L18",  lag_h = 18, ref = predTs_num - 18 * 3600),
  list(arm = "Parcel_L24",  lag_h = 24, ref = predTs_num - 24 * 3600),
  list(arm = "Parcel_L30",  lag_h = 30, ref = predTs_num - 30 * 3600),
  list(arm = "Parcel_L36",  lag_h = 36, ref = predTs_num - 36 * 3600),
  list(arm = "Parcel_model", lag_h = NA, ref = predTs_num - travel_tt * 3600)
)

# Compute label vectors per arm.
labels <- list()
for (s in arm_specs) {
  labels[[s$arm]] <- flowStateAsOf_vec(s$ref)
}
# Parcel_model: where travel_tt is NA, ref is NA -> findInterval(NA) returns NA.
# Handle: an NA ref means we cannot place it; treat as 'steady' (no current row resolvable).
# But spec assumes travel_time present. Check NA labels.
for (s in arm_specs) {
  na_lab <- sum(is.na(labels[[s$arm]]))
  if (na_lab > 0) {
    cat(sprintf("WARNING arm %s has %d NA labels (NA ref times) -> set to 'steady'\n", s$arm, na_lab))
    labels[[s$arm]][is.na(labels[[s$arm]])] <- "steady"
  }
}

# ---------------------------------------------------------------------------
# 4. Self-validation: label_match_rate (A_current vs logged flowState).
# ---------------------------------------------------------------------------
label_match_rate <- mean(labels[["A_current"]] == flowState_log)
cat(sprintf("label_match_rate (A_current vs logged): %.10f\n", label_match_rate))

# ---------------------------------------------------------------------------
# 5. Metrics helpers.
#    KEY = (flowBin, state). group means on rawResidual r.
# ---------------------------------------------------------------------------
r      <- rawResidual
N      <- length(r)
mean_r <- mean(r)
SS_total <- sum((r - mean_r)^2)

# group_means_for(state_vec): returns per-row group mean using KEY=(flowBin,state).
group_means_for <- function(state_vec) {
  key <- paste(flowBin, state_vec, sep = "")
  gm  <- ave(r, key, FUN = mean)
  gm
}

# Metric set for one labeling.
compute_metrics <- function(state_vec) {
  gm <- group_means_for(state_vec)
  SS_within <- sum((r - gm)^2)
  eta2_full <- 1 - SS_within / SS_total
  within_var <- SS_within / N

  # within_bin_state_eta2: hold flowBin fixed.
  # Sum over bins of SS_between_states / Sum over bins of SS_total_bin.
  # SS_between_states (within a bin) = SS_total_bin - SS_within_states_bin.
  # Equivalent: numerator = Sum_bin [ SS_total_bin - Sum_state (r - state_mean_in_bin)^2 ].
  # SS_total_bin = Sum_{rows in bin} (r - bin_mean)^2.
  bin_mean <- ave(r, flowBin, FUN = mean)
  SS_total_bin_total <- sum((r - bin_mean)^2)        # Sum_bin SS_total_bin
  # SS_within_states across bins == SS_within using KEY (bin,state) = SS_within already.
  SS_within_states_total <- SS_within
  SS_between_states_total <- SS_total_bin_total - SS_within_states_total
  within_bin_state_eta2 <- SS_between_states_total / SS_total_bin_total

  # Occupancy + rising mean resid.
  rising_n  <- sum(state_vec == "rising")
  steady_n  <- sum(state_vec == "steady")
  falling_n <- sum(state_vec == "falling")
  rising_mean_resid <- if (rising_n > 0) mean(r[state_vec == "rising"]) else NA_real_

  list(
    eta2_full = eta2_full,
    within_var = within_var,
    within_bin_state_eta2 = within_bin_state_eta2,
    rising_n = rising_n, steady_n = steady_n, falling_n = falling_n,
    rising_mean_resid = rising_mean_resid,
    group_means = gm,
    SS_within = SS_within
  )
}

metrics_by_arm <- list()
for (s in arm_specs) {
  metrics_by_arm[[s$arm]] <- compute_metrics(labels[[s$arm]])
}

# ---------------------------------------------------------------------------
# 6. Bootstrap (parcel arms vs A_current).
#    d_i = (r_i - gm_A(i))^2 - (r_i - gm_arm(i))^2.
#    Non-overlapping consecutive 48-row blocks, resample block indices w/ replacement,
#    B=1000, seed=20260617. boot_mean_d, boot_ci_lo/hi = 2.5/97.5 pctile of block-resampled mean(d).
#    delta_within_var = within_var[A_current] - within_var[arm].
#    NOTE: row order for blocks = the row order of the joined sample (file order of predictions),
#    which is chronological by predTs (verified ascending). Blocks are consecutive rows.
# ---------------------------------------------------------------------------
BLOCK <- 48
B <- 1000
SEED <- 20260617

gm_A <- metrics_by_arm[["A_current"]]$group_means
sq_A <- (r - gm_A)^2

# Partition rows into non-overlapping consecutive 48-row blocks.
# Last partial block (if N not divisible by 48) is kept as a shorter block.
n_blocks <- ceiling(N / BLOCK)
block_id <- ((seq_len(N) - 1L) %/% BLOCK) + 1L   # 1..n_blocks

# Precompute, per block: vector of d-values is needed; for resampling efficiency we
# precompute per-block sum(d) and length, then bootstrap mean = sum(selected sums)/sum(selected lens).
# This gives the resampled mean of d exactly (mean over all rows in the resampled blocks).

run_bootstrap <- function(arm_state_vec, arm_within_var) {
  gm_arm <- group_means_for(arm_state_vec)
  d <- sq_A - (r - gm_arm)^2
  delta_within_var <- metrics_by_arm[["A_current"]]$within_var - arm_within_var

  # Per-block sum(d) and length.
  block_sum_d <- tapply(d, block_id, sum)
  block_len   <- tapply(rep(1L, N), block_id, sum)
  block_sum_d <- as.numeric(block_sum_d)
  block_len   <- as.numeric(block_len)
  nb <- length(block_sum_d)

  boot_means <- numeric(B)
  for (b in seq_len(B)) {
    sel <- sample.int(nb, size = nb, replace = TRUE)
    boot_means[b] <- sum(block_sum_d[sel]) / sum(block_len[sel])
  }
  boot_mean_d <- mean(boot_means)
  ci <- quantile(boot_means, probs = c(0.025, 0.975), names = FALSE, type = 7)
  list(delta_within_var = delta_within_var,
       boot_mean_d = boot_mean_d,
       boot_ci_lo = ci[1], boot_ci_hi = ci[2])
}

# ---------------------------------------------------------------------------
# 7. Assemble output table (one row per arm). Seed reset before bootstrap loop
#    so the full sequence of parcel arms is deterministic.
# ---------------------------------------------------------------------------
set.seed(SEED)

rows <- list()
for (s in arm_specs) {
  m <- metrics_by_arm[[s$arm]]
  is_current <- s$arm == "A_current"
  lag_out <- if (is.na(s$lag_h)) NA_real_ else s$lag_h

  if (is_current) {
    boot <- list(delta_within_var = NA_real_, boot_mean_d = NA_real_,
                 boot_ci_lo = NA_real_, boot_ci_hi = NA_real_)
    lmr <- label_match_rate
  } else {
    boot <- run_bootstrap(labels[[s$arm]], m$within_var)
    lmr <- NA_real_
  }

  rows[[length(rows) + 1]] <- data.frame(
    arm = s$arm,
    lag_h = lag_out,
    n = N,
    eta2_full = m$eta2_full,
    within_var = m$within_var,
    within_bin_state_eta2 = m$within_bin_state_eta2,
    rising_n = m$rising_n,
    steady_n = m$steady_n,
    falling_n = m$falling_n,
    rising_mean_resid = m$rising_mean_resid,
    label_match_rate = lmr,
    delta_within_var = boot$delta_within_var,
    boot_mean_d = boot$boot_mean_d,
    boot_ci_lo = boot$boot_ci_lo,
    boot_ci_hi = boot$boot_ci_hi,
    stringsAsFactors = FALSE
  )
}

out <- do.call(rbind, rows)

# Write with full precision.
write.csv(out, OUT_PATH, row.names = FALSE)
cat(sprintf("Wrote %s\n", OUT_PATH))

# ---------------------------------------------------------------------------
# 8. Stdout summary.
# ---------------------------------------------------------------------------
cat("\n================ RESULTS (full precision) ================\n")
print(format(out, digits = 12, scientific = FALSE), row.names = FALSE)

cat("\n================ SUMMARY ================\n")
cat(sprintf("label_match_rate = %.6f\n", label_match_rate))
for (i in seq_len(nrow(out))) {
  if (out$arm[i] == "A_current") next
  reduces <- out$delta_within_var[i] > 0
  parcel_wins <- out$boot_ci_lo[i] > 0          # spec: "Parcel wins" <=> ci_lo > 0
  ci_excludes0 <- (out$boot_ci_lo[i] > 0) | (out$boot_ci_hi[i] < 0)
  cat(sprintf("%-13s lag=%s: within_var=%.4f delta_wv=%.4f (%s within_var); boot_mean_d=%.4f CI[%.4f, %.4f] %s; parcel_wins=%s\n",
              out$arm[i],
              ifelse(is.na(out$lag_h[i]), "model", as.character(out$lag_h[i])),
              out$within_var[i], out$delta_within_var[i],
              ifelse(reduces, "REDUCES", "increases"),
              out$boot_mean_d[i], out$boot_ci_lo[i], out$boot_ci_hi[i],
              ifelse(ci_excludes0, "[CI excludes 0]", "[CI includes 0]"),
              ifelse(parcel_wins, "YES", "NO")))
}

File: analysis/flow_state_step1.py
#!/usr/bin/env python3
"""
Step-1 Label-Only Diagnostic (Python implementation, blind dual-language).

Provenance: this is the generating script for `flow_state_step1_python.csv`.

Spec: analysis/flow-state-step1-diagnostic-spec.md
Question: does the parcel-aligned flow-state label explain more raw-residual
variance / yield tighter within-bin residuals than the current-T0 label, under
a lag-sensitivity sweep?

Inputs (read-only):
  - analysis/hourly_backtest_data_v361.csv  (hourly series, UTC)
  - analysis/ci_residuals_v361_multi.csv    (one row per validated prediction)

Implemented INDEPENDENTLY from the spec + raw data only. No R script, no
*_R.csv, no other agent output is read.

Deterministic; seed = 20260617 for the block bootstrap.
"""

import os
import numpy as np
import pandas as pd

# ----------------------------------------------------------------------------
# Paths
# ----------------------------------------------------------------------------
ANALYSIS_DIR = "/Users/sebjilke/Desktop/PotomacPulse/analysis"
HOURLY_CSV = os.path.join(ANALYSIS_DIR, "hourly_backtest_data_v361.csv")
CI_CSV = os.path.join(ANALYSIS_DIR, "ci_residuals_v361_multi.csv")
OUT_CSV = os.path.join(ANALYSIS_DIR, "flow_state_step1_python.csv")

SEED = 20260617
B = 1000
BLOCK = 48
HOUR = np.timedelta64(1, "h")
SIX_H = np.timedelta64(6, "h")

# ----------------------------------------------------------------------------
# Load series. Parse timestamps as UTC, drop blank por_now, sort ascending.
# ----------------------------------------------------------------------------
series = pd.read_csv(HOURLY_CSV)
series["ts"] = pd.to_datetime(series["timestamp"], utc=True, format="%Y-%m-%d %H:%M")
series = series.dropna(subset=["por_now"]).copy()
series = series.sort_values("ts").reset_index(drop=True)

# Numpy arrays for fast searchsorted. datetime64[ns] for the timestamps.
ts_arr = series["ts"].values.astype("datetime64[ns]")
por_arr = series["por_now"].values.astype(float)
N_SERIES = len(ts_arr)

# Map predTs -> travel_time_h (per-hour PoR->GF lag). Use the floored-hour join
# on the FULL hourly table (travel_time_h may exist on rows where por_now is
# blank, but the series we classify against is the por_now series; the lag value
# itself is a property of the prediction hour). Build lookup on full table.
full = pd.read_csv(HOURLY_CSV)
full["ts"] = pd.to_datetime(full["timestamp"], utc=True, format="%Y-%m-%d %H:%M")
tt_lookup = dict(zip(full["ts"].values.astype("datetime64[ns]"),
                     full["travel_time_h"].values.astype(float)))


# ----------------------------------------------------------------------------
# Classifier: flowStateAsOf(series, refTime)
#   hist = series rows with ts <= refTime
#   if len(hist) < 8: 'steady'
#   current = por_now of largest ts <= refTime
#   past    = por_now of largest ts <= refTime - 6h
#   if past is None: 'steady'
#   change = current - past ; thr = max(100, 0.02*current)
#   abs(change) >= thr -> 'rising' if change>0 else 'falling'; else 'steady'
# 6h is by TIME not index. Use searchsorted on the sorted ts array.
# ----------------------------------------------------------------------------
def flow_state_as_of(ref_time_ns):
    """ref_time_ns: numpy datetime64[ns]. Returns 'rising'/'steady'/'falling'."""
    # number of rows with ts <= refTime: index of first ts > refTime (side='right')
    n_hist = np.searchsorted(ts_arr, ref_time_ns, side="right")
    if n_hist < 8:
        return "steady"
    # current = nearest on-or-before refTime -> row at n_hist - 1
    cur_idx = n_hist - 1
    current = por_arr[cur_idx]
    # past = nearest on-or-before refTime - 6h
    past_ref = ref_time_ns - SIX_H
    n_past = np.searchsorted(ts_arr, past_ref, side="right")
    if n_past < 1:
        return "steady"  # no row on-or-before refTime-6h -> past is None
    past = por_arr[n_past - 1]
    change = current - past
    thr = max(100.0, 0.02 * current)
    if abs(change) >= thr:
        return "rising" if change > 0 else "falling"
    return "steady"


# Vectorized batch classifier over an array of refTimes (datetime64[ns]).
def classify_batch(ref_times_ns):
    """ref_times_ns: 1-D np.datetime64[ns] array (may contain NaT). Returns list of states."""
    out = np.empty(len(ref_times_ns), dtype=object)
    # n_hist for current
    n_hist = np.searchsorted(ts_arr, ref_times_ns, side="right")
    # n_past for refTime - 6h
    past_ref = ref_times_ns - SIX_H
    n_past = np.searchsorted(ts_arr, past_ref, side="right")
    for i in range(len(ref_times_ns)):
        rt = ref_times_ns[i]
        if np.isnat(rt):
            out[i] = "steady"  # undefined refTime -> default (see assumptions)
            continue
        nh = n_hist[i]
        if nh < 8:
            out[i] = "steady"
            continue
        current = por_arr[nh - 1]
        npst = n_past[i]
        if npst < 1:
            out[i] = "steady"
            continue
        past = por_arr[npst - 1]
        change = current - past
        thr = 100.0 if 0.02 * current < 100.0 else 0.02 * current
        if abs(change) >= thr:
            out[i] = "rising" if change > 0 else "falling"
        else:
            out[i] = "steady"
    return out


# ----------------------------------------------------------------------------
# Load predictions. Parse predTs as UTC.
# ----------------------------------------------------------------------------
ci = pd.read_csv(CI_CSV)
ci["predTs"] = pd.to_datetime(ci["predTs"], utc=True)
# Sort by predTs to define "consecutive" 48-row blocks (~2 days) by time.
ci = ci.sort_values("predTs").reset_index(drop=True)

predTs_ns = ci["predTs"].values.astype("datetime64[ns]")
r = ci["rawResidual"].values.astype(float)
flowBin = ci["flowBin"].values
logged_state = ci["flowState"].values
N = len(ci)

# travel_time_h per prediction row, via floored-hour join. predTs are already
# on the hour; lookup directly.
tt = np.array([tt_lookup.get(t, np.nan) for t in predTs_ns], dtype=float)

# ----------------------------------------------------------------------------
# Build arm labelings.
# ----------------------------------------------------------------------------
ARMS = []  # (arm_name, lag_h, refTimes_ns)

# A_current: refTime = predTs
ARMS.append(("A_current", 0.0, predTs_ns.copy()))

# Parcel_L{k}
for k in (6, 12, 18, 24, 30, 36):
    ref = predTs_ns - np.timedelta64(k, "h")
    ARMS.append((f"Parcel_L{k}", float(k), ref))

# Parcel_model: refTime = predTs - travel_time_h(predTs) h.
# travel_time may be fractional hours -> use float hours -> ns offset.
# If travel_time is NaN -> NaT refTime -> classify_batch returns 'steady'.
tt_offset_ns = np.where(np.isnan(tt), np.timedelta64("NaT"),
                        (tt * 3600.0 * 1e9).astype("timedelta64[ns]"))
# Build refTimes; NaT where tt is nan
ref_model = np.empty(N, dtype="datetime64[ns]")
for i in range(N):
    if np.isnan(tt[i]):
        ref_model[i] = np.datetime64("NaT")
    else:
        ref_model[i] = predTs_ns[i] - np.timedelta64(int(round(tt[i] * 3600.0 * 1e9)), "ns")
# Use the mean travel_time as the reported "lag_h" for Parcel_model (nominal).
ARMS.append(("Parcel_model", float(np.nanmean(tt)), ref_model))

# Classify all arms.
arm_states = {}
for name, lag, ref in ARMS:
    arm_states[name] = classify_batch(ref)

# ----------------------------------------------------------------------------
# label_match_rate: A_current reimplemented vs logged flowState.
# ----------------------------------------------------------------------------
a_current = arm_states["A_current"]
label_match_rate = float(np.mean(a_current == logged_state))


# ----------------------------------------------------------------------------
# Metric helpers. KEY = (flowBin, state).
# ----------------------------------------------------------------------------
def group_means(states):
    """Return array of group_mean for each row, where group = (flowBin, state)."""
    df = pd.DataFrame({"bin": flowBin, "state": states, "r": r})
    gm = df.groupby(["bin", "state"])["r"].transform("mean")
    return gm.values


def compute_deterministic(states):
    gm = group_means(states)
    grand = r.mean()
    ss_total = np.sum((r - grand) ** 2)
    ss_within = np.sum((r - gm) ** 2)
    eta2_full = 1.0 - ss_within / ss_total
    within_var = ss_within / N

    # within_bin_state_eta2: hold flowBin fixed.
    # Sum_bin SS_between_states / Sum_bin SS_total_bin
    df = pd.DataFrame({"bin": flowBin, "state": states, "r": r})
    num = 0.0  # Sum_bin SS_between_states
    den = 0.0  # Sum_bin SS_total_bin
    for b, gb in df.groupby("bin"):
        rb = gb["r"].values
        bin_mean = rb.mean()
        ss_total_bin = np.sum((rb - bin_mean) ** 2)
        # SS_between_states within this bin = Sum_state n_s*(state_mean - bin_mean)^2
        ss_between = 0.0
        for s, gs in gb.groupby("state"):
            rs = gs["r"].values
            ss_between += len(rs) * (rs.mean() - bin_mean) ** 2
        num += ss_between
        den += ss_total_bin
    within_bin_state_eta2 = num / den if den > 0 else np.nan

    # occupancy
    rising_n = int(np.sum(states == "rising"))
    steady_n = int(np.sum(states == "steady"))
    falling_n = int(np.sum(states == "falling"))
    rising_mask = states == "rising"
    rising_mean_resid = float(r[rising_mask].mean()) if rising_n > 0 else np.nan

    return dict(
        eta2_full=eta2_full,
        within_var=within_var,
        within_bin_state_eta2=within_bin_state_eta2,
        rising_n=rising_n,
        steady_n=steady_n,
        falling_n=falling_n,
        rising_mean_resid=rising_mean_resid,
        gm=gm,  # group means per row (for bootstrap d_i)
    )


det = {name: compute_deterministic(arm_states[name]) for name, _, _ in ARMS}

# ----------------------------------------------------------------------------
# Block bootstrap (per parcel arm vs A_current).
# d_i = (r_i - groupmean_A(i))^2 - (r_i - groupmean_arm(i))^2
# Non-overlapping consecutive 48-row blocks; resample block indices with
# replacement; B=1000; seed=20260617; report mean and 2.5/97.5 pctiles.
# delta_within_var = within_var[A_current] - within_var[arm]  (mean of d_i).
# ----------------------------------------------------------------------------
gm_A = det["A_current"]["gm"]
sq_A = (r - gm_A) ** 2

# Build non-overlapping blocks over the (predTs-sorted) sample.
# Block j covers rows [j*48 : (j+1)*48]. Last block may be shorter (kept).
n_blocks = int(np.ceil(N / BLOCK))
block_slices = []
for j in range(n_blocks):
    lo = j * BLOCK
    hi = min((j + 1) * BLOCK, N)
    block_slices.append((lo, hi))


def block_bootstrap(d):
    """d: per-row contribution array. Returns (mean_d, ci_lo, ci_hi)."""
    rng = np.random.default_rng(SEED)
    # Precompute per-block sums and counts of d for speed.
    block_sums = np.array([d[lo:hi].sum() for (lo, hi) in block_slices])
    block_counts = np.array([hi - lo for (lo, hi) in block_slices])
    boot = np.empty(B, dtype=float)
    for b in range(B):
        idx = rng.integers(0, n_blocks, size=n_blocks)  # resample block indices
        tot = block_sums[idx].sum()
        cnt = block_counts[idx].sum()
        boot[b] = tot / cnt
    mean_d = float(d.mean())
    ci_lo = float(np.percentile(boot, 2.5))
    ci_hi = float(np.percentile(boot, 97.5))
    return mean_d, ci_lo, ci_hi, boot.mean()


# ----------------------------------------------------------------------------
# Assemble output rows.
# ----------------------------------------------------------------------------
rows = []
for name, lag, _ in ARMS:
    d = det[name]
    row = dict(
        arm=name,
        lag_h=lag,
        n=N,
        eta2_full=d["eta2_full"],
        within_var=d["within_var"],
        within_bin_state_eta2=d["within_bin_state_eta2"],
        rising_n=d["rising_n"],
        steady_n=d["steady_n"],
        falling_n=d["falling_n"],
        rising_mean_resid=d["rising_mean_resid"],
        label_match_rate=label_match_rate if name == "A_current" else np.nan,
        delta_within_var=np.nan,
        boot_mean_d=np.nan,
        boot_ci_lo=np.nan,
        boot_ci_hi=np.nan,
    )
    if name != "A_current":
        gm_arm = d["gm"]
        sq_arm = (r - gm_arm) ** 2
        d_i = sq_A - sq_arm  # d_i; mean(d_i) == within_var[A] - within_var[arm]
        delta_within_var = det["A_current"]["within_var"] - d["within_var"]
        mean_d, ci_lo, ci_hi, boot_mean = block_bootstrap(d_i)
        row["delta_within_var"] = delta_within_var
        row["boot_mean_d"] = boot_mean  # bootstrap mean of resampled means
        row["boot_ci_lo"] = ci_lo
        row["boot_ci_hi"] = ci_hi
        # sanity: mean_d (observed) should equal delta_within_var
        row["_observed_mean_d"] = mean_d
    rows.append(row)

out = pd.DataFrame(rows)

# Column order per spec.
cols = ["arm", "lag_h", "n", "eta2_full", "within_var", "within_bin_state_eta2",
        "rising_n", "steady_n", "falling_n", "rising_mean_resid",
        "label_match_rate", "delta_within_var", "boot_mean_d",
        "boot_ci_lo", "boot_ci_hi"]
out_csv = out[cols].copy()
out_csv.to_csv(OUT_CSV, index=False)

# ----------------------------------------------------------------------------
# stdout summary.
# ----------------------------------------------------------------------------
pd.set_option("display.width", 200)
pd.set_option("display.max_columns", 50)
print("=== Step-1 Label-Only Diagnostic (Python) ===")
print(f"Joined sample N = {N}")
print(f"label_match_rate (A_current vs logged flowState) = {label_match_rate:.6f}")
print()
print(out[cols].to_string(index=False))
print()
print("--- Summary ---")
print(f"Match rate {label_match_rate:.4f}.")
for name, lag, _ in ARMS:
    if name == "A_current":
        continue
    rr = out[out["arm"] == name].iloc[0]
    reduces = rr["delta_within_var"] > 0
    wins = rr["boot_ci_lo"] > 0  # parcel wins iff CI strictly above 0
    if rr["boot_ci_hi"] < 0:
        ci_verdict = "EXCLUDES 0 on the LOSING side (parcel worse)"
    elif wins:
        ci_verdict = "EXCLUDES 0 (parcel wins)"
    else:
        ci_verdict = "straddles 0 (inconclusive)"
    print(f"{name}: delta_within_var={rr['delta_within_var']:.4f} "
          f"({'reduces' if reduces else 'does NOT reduce'} within_var vs A_current); "
          f"boot CI [{rr['boot_ci_lo']:.4f}, {rr['boot_ci_hi']:.4f}] {ci_verdict}.")
print()
print(f"Output written: {OUT_CSV}")
